nautilus/src/nautilus/tracing/tag/TagRecorder.cpp

```diff
@@ -9,133 +9,29 @@ namespace nautilus::tracing {
-	case 13:
-		return __builtin_return_address(13 + 1);
-	case 14:
-		return __builtin_return_address(14 + 1);
-	case 15:
-		return __builtin_return_address(15 + 1);
-	case 16:
-		return __builtin_return_address(16 + 1);
-	case 17:
-		return __builtin_return_address(17 + 1);
-	case 18:
-		return __builtin_return_address(18 + 1);
-	case 19:
-		return __builtin_return_address(19 + 1);
-	case 20:
-		return __builtin_return_address(20 + 1);
-	case 21:
-		return __builtin_return_address(21 + 1);
-	case 22:
-		return __builtin_return_address(22 + 1);
-	case 23:
-		return __builtin_return_address(23 + 1);
-	case 24:
-		return __builtin_return_address(24 + 1);
-	case 25:
-		return __builtin_return_address(25 + 1);
-	case 26:
-		return __builtin_return_address(26 + 1);
-	case 27:
-		return __builtin_return_address(27 + 1);
-	case 28:
-		return __builtin_return_address(28 + 1);
-	case 29:
-		return __builtin_return_address(29 + 1);
-	case 30:
-		return __builtin_return_address(30 + 1);
-	case 31:
-		return __builtin_return_address(31 + 1);
-	case 32:
-		return __builtin_return_address(32 + 1);
-	case 33:
-		return __builtin_return_address(33 + 1);
-	case 34:
-		return __builtin_return_address(34 + 1);
-	case 35:
-		return __builtin_return_address(35 + 1);
-	case 36:
-		return __builtin_return_address(36 + 1);
-	case 37:
-		return __builtin_return_address(37 + 1);
-	case 38:
-		return __builtin_return_address(38 + 1);
-	case 39:
-		return __builtin_return_address(39 + 1);
-	case 40:
-		return __builtin_return_address(40 + 1);
-	case 41:
-		return __builtin_return_address(41 + 1);
-	case 42:
-		return __builtin_return_address(42 + 1);
-	case 43:
-		return __builtin_return_address(43 + 1);
-	case 44:
-		return __builtin_return_address(44 + 1);
-	case 45:
-		return __builtin_return_address(45 + 1);
-	case 46:
-		return __builtin_return_address(46 + 1);
-	case 47:
-		return __builtin_return_address(47 + 1);
-	case 48:
-		return __builtin_return_address(48 + 1);
-	case 49:
-		return __builtin_return_address(49 + 1);
-	case 50:
-		return __builtin_return_address(50 + 1);
-	case 51:
-		return __builtin_return_address(51 + 1);
-	case 52:
-		return __builtin_return_address(52 + 1);
-	case 53:
-		return __builtin_return_address(53 + 1);
-	case 54:
-		return __builtin_return_address(54 + 1);
-	case 55:
-		return __builtin_return_address(55 + 1);
-	case 56:
-		return __builtin_return_address(56 + 1);
-	case 57:
-		return __builtin_return_address(57 + 1);
-	case 58:
-		return __builtin_return_address(58 + 1);
-	case 59:
-		return __builtin_return_address(59 + 1);
-	case 60:
-		return __builtin_return_address(60 + 1);
-	default:
-		return nullptr;
+template<size_t StackSize>
+struct ReturnAddressHelper {
+	template<size_t Depth = 1>
+	static void* get_addr(size_t index){
+		if(index == StackSize){
+			return __builtin_return_address(StackSize + Depth);
+		}else{
+			return ReturnAddressHelper<StackSize-1>::template get_addr<Depth + 1>(index);
```

Comment:
hm is this recursion unrolled?

Reply:
Nope it is not. Is this a problem?

Reply:
This is also the reason why we have to count the depth

Reply:
I just tried it here:
https://godbolt.org/z/64GcP9jrK
It looks like that its indeed generating function calls. 
As each function call is basically a new stack frame this prevents us to get basically the correct stack frame addresses. 
In theory the compiler can inline all these calls, but this is not always guaranteed and may only work with O3. 
We could use __attribute__((musttail)). https://stackoverflow.com/questions/4785066/is-it-possible-to-force-tail-call-optimization-on-gcc-clang
But this is not supported in gcc... :(